turbopack/crates/turbo-tasks-backend/src/kv_backing_storage.rs

```diff
@@ -797,14 +914,119 @@ fn serialize(task: TaskId, data: &Vec<CachedDataItem>) -> Result<SmallVec<[u8; 1
+        if is_new {
+            batch.put(
+                KeySpace::StringInternMap,
+                WriteBuffer::Borrowed(s.as_bytes()),
+                WriteBuffer::Borrowed(&global_id.to_le_bytes()),
+            )?;
+            batch.put(
+                KeySpace::ReverseStringInternMap,
+                WriteBuffer::Borrowed(IntKey::new(global_id).as_ref()),
+                WriteBuffer::Borrowed(s.as_bytes()),
+            )?;
+        }
+
+        global_ids.push(global_id);
+    }
+
+    Ok(LocalIdToGlobalId::from(global_ids))
+}
+/// Store the strings in the database and return the global ids.
+fn save_strings_concurrent<'a>(
+    batch: &impl ConcurrentWriteBatch<'a>,
+    strings: &RcStrToLocalId,
+) -> Result<LocalIdToGlobalId> {
+    let mut global_ids = Vec::new();
+    for s in strings.iter() {
+        let (global_id, is_new) = get_string_id(batch, s)?;
+        if is_new {
+            batch.put(
+                KeySpace::StringInternMap,
+                WriteBuffer::Borrowed(s.as_bytes()),
+                WriteBuffer::Borrowed(&global_id.to_le_bytes()),
+            )?;
+            batch.put(
+                KeySpace::ReverseStringInternMap,
+                WriteBuffer::Borrowed(IntKey::new(global_id).as_ref()),
+                WriteBuffer::Borrowed(s.as_bytes()),
+            )?;
+        }
+
+        global_ids.push(global_id);
+    }
+
+    Ok(LocalIdToGlobalId::from(global_ids))
+}
+
+static STRING_INTERN_ID: OnceLock<AtomicU32> = OnceLock::new();
+static STRING_CACHE: LazyLock<Mutex<FxHashMap<u32, RcStr>>> =
+    LazyLock::new(|| Mutex::new(FxHashMap::default()));
+static STRING_TO_ID_CACHE: LazyLock<FxDashMap<RcStr, u32>> = LazyLock::new(FxDashMap::default);
+
+/// Returns `(global_id, is_new)`
+fn get_string_id<'a>(batch: &impl BaseWriteBatch<'a>, s: &RcStr) -> Result<(u32, bool)> {
+    // First check the synchronization cache
+    if let Some(global_id) = STRING_TO_ID_CACHE.get(s) {
+        return Ok((*global_id, false));
+    }
+
+    // Check the database
+    let original = batch.get(KeySpace::StringInternMap, s.as_bytes())?;
+
+    if let Some(bytes) = original {
+        let global_id = as_u32(bytes)?;
+        // Cache the result for future lookups
+        STRING_TO_ID_CACHE.insert(s.clone(), global_id);
+        return Ok((global_id, false));
+    }
+
+    // Allocate a new ID
+    let global_id = STRING_INTERN_ID
+        .get_or_try_init(|| {
+            let Some(bytes) = batch.get(
+                KeySpace::Infra,
+                IntKey::new(META_KEY_NEXT_STRING_ID).as_ref(),
+            )?
+            else {
+                return anyhow::Ok(AtomicU32::new(0));
+            };
+
+            let latest_id = as_u32(bytes)?;
+
+            Ok(AtomicU32::new(latest_id))
+        })?
+        .fetch_add(1, Ordering::Relaxed);
+
+    // Cache the new assignment to prevent duplicates within the same batch
+    STRING_TO_ID_CACHE.insert(s.clone(), global_id);
```

Comment:
Again there is a race condition here.

Use STRING_TO_ID_CACHE.entry() and return `Ok((existing_id, false))` if there is already an entry for this id